Thumbnail

mokaddem opened the pull request:
### :warning: WARNING :warning:
- This PR introduces a new dependency ``php-gd``
- All install files should have been updated accordingly
- Also related to https://github.com/MISP/misp-packer/pull/15

## Features

- Asynchronous pictures loading in ``event view``
- Improved Image modal view allowing large pictures
- Added ``viewPicture`` endpoint
- Integrated picture re-sampling allowing to resize picture on-demand
- Added support in diagnostic

mokaddem committed:
chg: [Attribute] Image thumbnail and view image

mokaddem committed:
new: [Attribute] Added possibility to view and resize images. Added
php-gd dependency!

mokaddem committed:
chg: clean-up

mokaddem committed:
chg: cleanup 2

mokaddem committed:
chg: [Attribute] Added loading icon when downloading pictures

mokaddem committed:
chg: [viewPicture] Adaptative behavior if php-gd not loaded

mokaddem committed:
chg: [diagnostic] Added GD in PHP extension

mokaddem committed:
chg: [INSTALL] Added GD dependency

mokaddem committed:
fix: [viewField] Reversed condition (initially set for testing)

mokaddem committed:
fix: [ACL] Whitelisted `viewPicture`

adulau commented:
This pull request **fixes 1 alert** when merging 749c9ea54416280d8e6c73faa802c8078b5a99c5 into 695771eb44ccb650b35bd8a973972b7c2996cd99 - [view on LGTM.com](https://lgtm.com/projects/g/MISP/MISP/rev/pr-332d4c6741260cec5f0d77c10728a3d9eeb581c3)

**fixed alerts:**

* 1 for Missing variable declaration

---

*Comment posted by [LGTM.com](https://lgtm.com)*

iglocska commented on app/Controller/AttributesController.php:
```diff
@@ -1189,7 +1189,85 @@ public function view($id)
+            throw new MethodNotAllowedException('Invalid attribute');
+        }
+        $attribute = $attribute[0];
+        if ('attachment' != $attribute['Attribute']['type'] && 'malware-sample' != $attribute['Attribute']['type'] ) {
```
Why not include this in the conditions already? Also, malware sample should NOT be a balid type for thumbnails.

mokaddem replied:
Copy pasta master... :1st_place_medal: 
You are totally right!

iglocska commented on app/View/Elements/Events/View/value_field.ctp:
```diff
@@ -2,10 +2,16 @@
-      echo '<img class="screenshot screenshot-collapsed useCursorPointer" src="' . $uri . '" title="' . h($object['value']) . '" />';
+        if (extension_loaded('gd')) {
+            $img = '<it class="fa fa-spin fa-spinner" style="font-size: large; left: 50%; top: 50%;"></it>';
+            $img .= '<img class="screenshot screenshot-collapsed useCursorPointer img-rounded" src="' . $baseurl . '/attributes/viewPicture/' . h($object['id']) . '/1' . '" title="' . h($object['value']) . '" onload="$(this).parent().find(\'.fa-spinner\').remove();"/>';
```
This looks great, however, keep in mind that the image is still loaded, b64 encoded and passed to the view - it just isn't transferred to the user. Perhaps also change the bahaviour on the controller side (EventsController->view() / EventsController->__viewUI() / EventsController->viewEventAttributes() - or whatever it's called).

mokaddem replied:
Sure! Especially that this PR aims to improve the perf...
will have a look.

mokaddem committed:
Merge branch '2.4' of github.com:MISP/MISP into thumbnail

mokaddem committed:
chg: [Attribute] Implemented Iglocska review
- Do not pass image data to the view anymore
- Improved conditions
- Added light support of gif

mokaddem committed:
chg: [viewPicture] Echo base64decoded data with GIF image type as php-gd does not support
animated gif.

adulau commented:
This pull request **fixes 1 alert** when merging 2ac6d2434e951a33a630517a900d858772793b65 into 43f0010cab9776d5bb447dc230f488b8229ffa6d - [view on LGTM.com](https://lgtm.com/projects/g/MISP/MISP/rev/pr-883a843552d8eb99d2880087cd0a09858ea820ce)

**fixed alerts:**

* 1 for Missing variable declaration

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mokaddem committed:
fix: [picturePreview] Don't show the fullscreen button if php-gd is not
installed.

mokaddem committed:
chg: [INSTALL] Do not touch the auto-generated installation file anymore

mokaddem committed:
fix: [thumbnail] picture preview was not showing up all the time

mokaddem commented:
Recently fixed a bug that was occasionally preventing to show the picture if ``php-gd`` wasn't installed.
Could someone give it another try? @iglocska this time? :)

iglocska commented:
Works for me with and without the extension, awesome job!
